x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/ILMHistoryTests.java

```diff
@@ -0,0 +1,104 @@
+public class ILMHistoryTests extends ESIntegTestCase {
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        Settings.Builder settings = Settings.builder().put(super.nodeSettings(nodeOrdinal));
+        settings.put(XPackSettings.MACHINE_LEARNING_ENABLED.getKey(), false);
+        settings.put(XPackSettings.SECURITY_ENABLED.getKey(), false);
+        settings.put(XPackSettings.WATCHER_ENABLED.getKey(), false);
+        settings.put(XPackSettings.GRAPH_ENABLED.getKey(), false);
+        settings.put(LifecycleSettings.LIFECYCLE_POLL_INTERVAL, "1s");
+        settings.put(LifecycleSettings.SLM_HISTORY_INDEX_ENABLED_SETTING.getKey(), false);
+        return settings.build();
+    }
+
+    @Override
+    protected boolean ignoreExternalCluster() {
+        return true;
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Arrays.asList(LocalStateCompositeXPackPlugin.class, IndexLifecycle.class);
+    }
+
+    private void putTestPolicy() throws InterruptedException, java.util.concurrent.ExecutionException {
+        LifecyclePolicy lifecyclePolicy = randomTimeseriesLifecyclePolicy("test");
+        PutLifecycleAction.Request putLifecycleRequest = new PutLifecycleAction.Request(lifecyclePolicy);
+        PutLifecycleAction.Response putLifecycleResponse = client().execute(PutLifecycleAction.INSTANCE, putLifecycleRequest).get();
+        assertAcked(putLifecycleResponse);
+    }
+
+    public void testIlmHistoryIndexCanRollover() throws Exception {
+        putTestPolicy();
+        Settings settings = Settings.builder().put(indexSettings()).put(SETTING_NUMBER_OF_SHARDS, 1)
+            .put(SETTING_NUMBER_OF_REPLICAS, 0).put(LifecycleSettings.LIFECYCLE_NAME, "test").build();
+        CreateIndexResponse res = client().admin().indices().prepareCreate("test").setSettings(settings).get();
+        assertTrue(res.isAcknowledged());
+
+        String firstIndex = ILMHistoryStore.ILM_HISTORY_INDEX_PREFIX + "000001";
+        String secondIndex = ILMHistoryStore.ILM_HISTORY_INDEX_PREFIX + "000002";
+
+        assertBusy(() -> {
+            try {
+                GetIndexResponse getIndexResponse = client().admin().indices().prepareGetIndex().setIndices(firstIndex).get();
+                assertThat(getIndexResponse.getIndices(), arrayContaining(firstIndex));
+            } catch (Exception e) {
+                fail(e.getMessage());
+            }
+        });
+
+        assertTrue(client().execute(StopILMAction.INSTANCE, new StopILMRequest()).actionGet().isAcknowledged());
+
+        //wait for all history items to index to avoid waiting for timeout in ILMHistoryStore beforeBulk
+        assertBusy(() -> {
+            SearchResponse search = client().prepareSearch(firstIndex).setQuery(matchQuery("index", firstIndex)).setSize(0).get();
+            assertThat(search.getHits().getTotalHits().value, greaterThan(0L));
```

Comment:
should we avoid the randomisation of the ILM policy and use an empty hot phase only ILM policy? that way we can assert exactly how many items must be inserted before moving on with the test tear down. I _think_ there might be another race condition possible otherwise as the stop ILM call is not a hard stop but has to be detected and depending on what step we're in we might execute more steps (see https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleService.java#L133 )

Reply:
You may be right there is still flakiness there. I've simplified policy, wait for exact number of messages to be indexed and wait for ILM to be stopped